File: cpp/data_structures/tree/binary_tree/binary_tree.cpp
#include <alg_std.hpp>
#include <cassert>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode(int v, TreeNode* l = nullptr, TreeNode* r = nullptr) : val(v), left(l), right(r) {}
};

void preorder(TreeNode* r, vector<int>& out) {
    if (!r) return;
    out.push_back(r->val);
    preorder(r->left, out);
    preorder(r->right, out);
}

vector<int> inorder_iter(TreeNode* root) {
    vector<int> out;
    stack<TreeNode*> st;
    TreeNode* cur = root;
    while (!st.empty() || cur) {
        while (cur) {
            st.push(cur);
            cur = cur->left;
        }
        cur = st.top();
        st.pop();
        out.push_back(cur->val);
        cur = cur->right;
    }
    return out;
}

vector<vector<int>> levelOrder(TreeNode* root) {
    vector<vector<int>> res;
    if (!root) return res;
    queue<TreeNode*> q;
    q.push(root);
    while (!q.empty()) {
        int sz = (int)q.size();
        vector<int> level;
        for (int i = 0; i < sz; ++i) {
            TreeNode* n = q.front();
            q.pop();
            level.push_back(n->val);
            if (n->left) q.push(n->left);
            if (n->right) q.push(n->right);
        }
        res.push_back(move(level));
    }
    return res;
}

int main() {
    TreeNode n4(4), n2(2, &n4, nullptr), n3(3), root(1, &n2, &n3);
    vector<int> pre;
    preorder(&root, pre);
    assert(pre == vector<int>({1, 2, 4, 3}));
    assert(inorder_iter(&root) == vector<int>({4, 2, 1, 3}));
    auto lv = levelOrder(&root);
    assert(lv.size() == 3);
    cout << "BinaryTree OK" << endl;
    return 0;
}

File: python/data_structures/tree/binary_tree/binary_tree.py
"""二叉树：四种遍历（递归 + 迭代）。"""

from __future__ import annotations
from collections import deque


class TreeNode:
    __slots__ = ("val", "left", "right")

    def __init__(self, val: int, left: TreeNode | None = None, right: TreeNode | None = None) -> None:
        self.val = val
        self.left = left
        self.right = right


def preorder_recursive(root: TreeNode | None) -> list[int]:
    out: list[int] = []

    def dfs(n: TreeNode | None) -> None:
        if n is None:
            return
        out.append(n.val)
        dfs(n.left)
        dfs(n.right)

    dfs(root)
    return out


def inorder_iterative(root: TreeNode | None) -> list[int]:
    out: list[int] = []
    st: list[TreeNode] = []
    cur: TreeNode | None = root
    while st or cur is not None:
        while cur is not None:
            st.append(cur)
            cur = cur.left
        cur = st.pop()
        out.append(cur.val)
        cur = cur.right
    return out


def postorder_recursive(root: TreeNode | None) -> list[int]:
    out: list[int] = []

    def dfs(n: TreeNode | None) -> None:
        if n is None:
            return
        dfs(n.left)
        dfs(n.right)
        out.append(n.val)

    dfs(root)
    return out


def level_order(root: TreeNode | None) -> list[list[int]]:
    if root is None:
        return []
    res: list[list[int]] = []
    q: deque[TreeNode] = deque([root])
    while q:
        level: list[int] = []
        for _ in range(len(q)):
            n = q.popleft()
            level.append(n.val)
            if n.left is not None:
                q.append(n.left)
            if n.right is not None:
                q.append(n.right)
        res.append(level)
    return res


if __name__ == "__main__":
    #       1
    #      / \
    #     2   3
    #    /
    #   4
    root = TreeNode(1, TreeNode(2, TreeNode(4)), TreeNode(3))
    assert preorder_recursive(root) == [1, 2, 4, 3]
    assert inorder_iterative(root) == [4, 2, 1, 3]
    assert postorder_recursive(root) == [4, 2, 3, 1]
    assert level_order(root) == [[1], [2, 3], [4]]
    assert level_order(None) == []
    assert preorder_recursive(None) == []
    print("BinaryTree OK")
